p2p/module.go

```diff
@@ -124,31 +153,35 @@ func (m *p2pModule) Start() error {
 			telemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_DESCRIPTION,
 		)
 
-	go func() {
-		for {
-			data, err := m.listener.ReadAll()
-			if err != nil {
-				m.logger.Error().Err(err).Msg("Error reading data from connection")
-				continue
-			}
-			go m.handleNetworkMessage(data)
+	// Return early if host has already been started (e.g. via `WithHostOption`)
+	if m.host == nil {
+		if err = m.setupHost(); err != nil {
+			return fmt.Errorf("setting up libp2pHost: %w", err)
 		}
-	}()
+	}
+
+	if err := m.setupNetwork(); err != nil {
+		return fmt.Errorf("setting up network: %w", err)
+	}
+
+	// Don't handle incoming streams in client debug mode.
+	if !m.isClientDebugMode() {
+		m.host.SetStreamHandler(protocol.PoktProtocolID, m.handleStream)
+	}
 
 	m.GetBus().
 		GetTelemetryModule().
 		GetTimeSeriesAgent().
 		CounterIncrement(telemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_NAME)
-
 	return nil
 }
 
 func (m *p2pModule) Stop() error {
-	m.logger.Info().Msg("Stopping network module")
-	if err := m.listener.Close(); err != nil {
-		return err
-	}
-	return nil
+	err := m.host.Close()
+
+	// Don't reuse closed host, `#Start()` will re-create.
+	m.host = nil
```

Comment:
There's a bug here :roll_eyes:

If the module was created with a `WithHost()` option, `#Start()` will ignore if a P2P module is stopped and restarted (which is what the regression test for this will look like).

Reply:
Can't we just do a thing where change the stop logic if there's a `WithHost` option provided?

Reply:
Unfortunately it doesn't seem quite that simple, and I'm realizing that the solution I described is still insufficient: 

1. Options are indistinguishable from one another and opaque with respect to what they modify on whatever gets passed to the option function.
1. A libp2p host can't be stopped, it can only be destroyed, after which it would need to be recreated. This is especially problematic in a "bring your own host" situation as we have no way to re-create that host.

TLDR; we should return an error in the case where a module `WithHost` is restarted. Apologies for the back and forth.